### ISSUE TITLE
Add boost data.

### ISSUE THREAD

**Issue by enzanki-ars** [enzanki-ars/rocket-league-minimap-generator#15]
There is boost data in the replay file, but it seem hard to parse.  Every so often the value is explicitly updated, but does not seem to be updated on use or pickup.  

In terms of the bounty: 
- I don't have much, and I don't value crates, but there seems to be a weird interest in them.  I have a very small amount of them, but they are up for grabs if desired.  
- I will award the bounty if someone is able to document how the boost data updates work, and is either able to code the parser, or to help point me in the right direction.  Preferably, the boost data should be as accurate as possible per frame, including pickup and usage of the boost.

**Comment by jjbott** [comment]
I'm about 4% sure that:
- You need to know the value of every boost pickup
- You need to know the rate of boost consumption

Then you can calculate the boost at any point in time based on how when it is picked up and when it is used. You can use the values in the replay to double check your accuracy.

**Comment by jeromepl** [comment]
I have worked on something similar from Rattletrap replay files in the past and did indeed do exactly what @jjbott listed. I had found that the rate at which boost is consumed is almost exactly 33.33% per second (or 100 boost to 0 in 3 seconds, or 85/255 per second if stored in a byte).

**Comment by enzanki-ars** [comment]
Thank you for the tips @jjbott and @jeromepl.  The only problem is that either I am missing something obvious, or I am missing the value of boost pickups, who picks it up, and when usage starts/stops.  I see something referred to as "BoostPickup", but I don't see a way to tie it to a player...

There is something obvious I am missing, but I just can't figure it out.

**Comment by jjbott** [comment]
The value of the boost pickups isn't in the replays. Somehow we'd need to figure out that, for example, "UtopiaStadium_Snow_p.TheWorld:PersistentLevel.VehiclePickup_Boost_TA_21" is 100 boost (maybe, I made that value up).

Pretty sure "TAGame.VehiclePickup_TA:ReplicatedPickupData" shows when a boost was picked up, and by who. The number in the data is the id if the car that picked it up. I should make that more obvious in the json...

Boost on/off is "TAGame.CarComponent_TA:ReplicatedActive". If the value is odd, boost is on.  I've got that logic here, although I suspect that's currently dead code for some reason: https://github.com/jjbott/RocketLeagueReplayParser/blob/9314b551550a3ed24cd3cd3835eb028a7534358a/RocketLeagueReplayParser/Serializers/ActorStateJsonConverter.cs#L72

**Comment by jeromepl** [comment]
I would need to do more research on how Rattletrap and RocketLeagueReplayParser compare, but when using Rattletrap, I was able to attach a `Archetypes.CarComponents.CarComponent_Boost` actor (along with its actor ID) to a Player. From there, when a replication matched the boost actor ID and that the value name was `TAGame.CarComponent_Boost_TA:ReplicatedBoostAmount`, the byte value can be used to update the boost amount. I had found that this happens on boost pickup.
On top of that, boost needs to decrease based on the consumption rate, and for that you need to know if a player is currently boosting. As @jjbott said, this can easily be found by checking if `TAGame.CarComponent_TA:ReplicatedActive` is odd. Once you got that, you can decrease the boost amount by the consumption rate multiplied by the delta time.

The portion of code I had for this was the following (inside a loop going over the replications in each frame of the replay file):
```javascript
if (player.boostActorID && player.boostActorID === replication.actor_id.value) {
    replicationValue.forEach((updatedReplicationValue) => {
        if (updatedReplicationValue.name === 'TAGame.CarComponent_Boost_TA:ReplicatedBoostAmount') {
            player.setBoostValue(updatedReplicationValue.value.byte_attribute_value);
        } else {
            // Since boost value is not updated in every frame while a player is boosting, manually update its value
            // based on whether or not the player is currently boosting.
            player.updateBoostValue(delta);
        }
        
        if (updatedReplicationValue.name === 'TAGame.CarComponent_TA:ReplicatedActive') {
            // Odd values means the player started boosting, even values means he/she stopped boosting or ran out of boost
            player.boosting = updatedReplicationValue.value.byte_attribute_value % 2 === 1;
        }
    });
}
```

Inside `Player.updateBoostValue(deltaTime)` I have the following (where `BOOST_CONSUMPTION_RATE = 85`):
```javascript
updateBoostValue(delta) {
    if (this.boosting) {
        this.boostValue -= delta * BOOST_CONSUMPTION_RATE;
        this.boostValue = Math.max(0, this.boostValue);
    }
}
```

**Comment by enzanki-ars** [comment]
@jeromepl Thanks so much.  That should help a bunch, and I should be able to translate that very easily as Rattletrap and RocketLeagueReplayParser are very similar.

I have some other things that I need to fix first, but boost data should be added in very shortly then.  Thank you so much for your assistance.

**Comment by enzanki-ars** [comment]
Based on what I see here, I will remove the bounty tag.  

@jeromepl feel free to PM on Reddit at /u/enzanki_ars or on twitter @enzanki_ars if you are interested in a couple of crates.  I know that they are useless without keys, but I don't play much competitive/online to get drops, so I have absolutely nothing else.

**Comment by jeromepl** [comment]
Glad I could help! Don't worry about the crates, I don't care about them too much either ;)
If I find the time later next week, I might take a deeper look at this project and try contributing! Keep up the good work!

**Comment by enzanki-ars** [comment]
Thanks again for the help @jjbott and @jeromepl! I was able to add in the boost data, and add a boost comparison render.
